rolling-shutter/collator/batchhandler/sequencer/statecache.go

```diff
@@ -27,34 +30,104 @@ type EthClient interface {
 	NonceAt(context.Context, common.Address, *big.Int) (uint64, error)
 }
 
-func NewChainBatchCache(client EthClient, atBlockNumber *big.Int) *ChainBatchCache {
-	return &ChainBatchCache{
+type countedMux struct {
+	mux *sync.RWMutex
+	cnt uint32
```

Comment:
Maybe clarify that we use the same counter for `Lock/Unlock` and `RLock/RUnlock`, at least to me this was confusing (and not self-evident that it's ok to do so).

Reply:
Will document it.
I also was not sure initially, but this counter functions as a reference counter to the mutex object.
It tracks general usage of the mutex in order to not keep dangling objects hanging in the map
indefinitely, and to instantiate a new one when the counter went to 0 before.
The object itself decides about read/write access separation.